### PR TITLE
BUG: User-facing AssertionError with add column to SparseDataFrame

### LINKED ISSUE #25484: BUG: User-facing AssertionError with add column to SparseDataFrame
see also #5289

#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
data = {'A': [np.nan, np.nan, np.nan, 0, 1, 2, 3, 4, 5, 6],
        'B': [0, 1, 2, np.nan, np.nan, np.nan, 3, 4, 5, 6],
        'C': np.arange(10, dtype=np.float64),
        'D': [0, 1, 2, 3, 4, 5, np.nan, np.nan, np.nan, np.nan]}
dates = pd.bdate_range('1/1/2011', periods=10)
float_frame = pd.SparseDataFrame(data, index=dates, default_kind='block')
N = len(float_frame)
float_frame['foo'] = np.random.randn(N - 1)
```
#### Problem description

```python-traceback
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-8-0a4f051f0d7a> in <module>
      8 float_frame = pd.SparseDataFrame(data, index=dates, default_kind='block')
      9 N = len(float_frame)
---> 10 float_frame['foo'] = np.random.randn(N - 1)

~\Anaconda3\lib\site-packages\pandas\core\frame.py in __setitem__(self, key, value)
   3117         else:
   3118             # set column
-> 3119             self._set_item(key, value)
   3120 
   3121     def _setitem_slice(self, key, value):

~\Anaconda3\lib\site-packages\pandas\core\frame.py in _set_item(self, key, value)
   3192 
   3193         self._ensure_valid_index(value)
-> 3194         value = self._sanitize_column(key, value)
   3195         NDFrame._set_item(self, key, value)
   3196 

~\Anaconda3\lib\site-packages\pandas\core\sparse\frame.py in _sanitize_column(self, key, value, **kwargs)
    418             else:
    419                 if len(value) != len(self.index):
--> 420                     raise AssertionError('Length of values does not match '
    421                                          'length of index')
    422                 clean = sp_maker(value)

AssertionError: Length of values does not match length of index
```

#### Expected Output

```python-traceback
ValueError: Length of values does not match length of index
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: 4.0.2
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [ ] closes #25484 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
Good start!  Definitely will need to add a test for this.

**Comment by simonjayhawkins** [comment]
> Good start! Definitely will need to add a test for this.

one of the changes will be hitting this test (updated to catch error type in #25483)

https://github.com/pandas-dev/pandas/blob/b99f8bc3ce4b0d5aa71e2caf8db5e6dc4dbc42de/pandas/tests/sparse/frame/test_frame.py#L553-L555

File: pandas/tests/sparse/frame/test_frame.py [hunk 1]
```diff
@@ -156,6 +156,15 @@ def test_constructor_ndarray(self, float_frame):
...
 
+    def test_construct_unmatched_sparsedataframe(self, float_frame):
+        # GH25484
+        dates = pd.bdate_range('1/1/2011', periods=10)
+        sp = SparseDataFrame(float_frame.values, index=dates)
+        N = len(sp)
+        msg = 'Length of values does not match length of index'
+        with pytest.raises(ValueError, match=msg):
+            sp['foo'] = np.random.randn(N-1)
+
```
Comment (simonjayhawkins) [inline pandas/tests/sparse/frame/test_frame.py]: may not need an additional test (https://github.com/pandas-dev/pandas/pull/25503#issuecomment-468906150)

the code sample in #25484 replicates the setup of the existing test. so maybe able to alter existing test to hit both changes you've made to `pandas\core\sparse\frame.py in _sanitize_column`

Comment (charlesdong1991) [reply pandas/tests/sparse/frame/test_frame.py]: ahh, i see! thanks! @simonjayhawkins

**Comment by jreback** [comment]
can you merge master and resolve conflicts. ping on green.

**Comment by jreback** [comment]
can you merge master & add a whatsnew note for this. ping on green.

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -227,7 +227,7 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
-
+- Bug in `SparseDataFrame` when adding a column in which the length of values does not match length of index, ``AssertionError`` is raised instead of raising ``ValueError`` (:issue:`25484`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you make a Sparse sub-section and move this there

**Review by jreback — CHANGES_REQUESTED** [review]
doc-change, otherwise lgtm. pls merge master. ping on green.

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (charlesdong1991) [reply doc/source/whatsnew/v0.25.0.rst]: okay, sparse section is already there, moved! thanks @jreback

**Comment by charlesdong1991** [comment]
@jreback i merged the master and made the doc change, pls take a look, thanks

**Comment by charlesdong1991** [comment]
@jreback i just merged the master and resolved the conflict, pls take a look if you have some time. thanks!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @charlesdong1991

**Pull request merged**